src/python/pants/option/subsystem.py

```diff
@@ -67,6 +68,18 @@ class Subsystem(metaclass=_SubsystemMeta):
     options_scope: str
     help: ClassVar[str | Callable[[], str]]
 
+    @classproperty
+    def help_extended(cls) -> str:
+        """
+        Help text to be used in `./pants help`.
+        Subclasses may override this to add more information,
+        but by default, the text generated by `help` is returned unchanged.
+        """
+        if isinstance(cls.help, str):
+            return cls.help
+
+        return cls.help()
```

Comment:
The challenge here is that subclasses of `PythonToolRequirementBase` will override `help` as well and make it a string (as you said). 

So I didn't want to interfere with any of the current functionality and cause breakage to any existing help text. It seemed simpler to just add in a new method and invoke that instead of messing with the existing structure.

Reply:
Fair enough!

Reply:
Hm, sorry to revisit this, but I'm a bit nervous about this approach, as it seems like we're pushing something that's somewhat of an implementation detail of `PythonToolRequirementBase` / the Python subsystems up to the top-level subsystem concept, side stepping the normal tools for this in Python (i.e. subclassing & overrides). All of these becomes available to third-party plugins, so exposing it has future-design/backwards-compatibility concerns (i.e. if we land this and then reconsider, we might have to go through deprecation cycles etc.).

I can think of a few options:

1. keep this approach, with a new method on the `Subsystem` class which lets "leaf" subclasses just set `help` while still giving superclasses flexibility for manipulating it
2. a new field manipulated by the Python tool helpers, that Python tool subclasses need to set
3. a class method that Python tool subclasses can call when formatting their own help text
4. omit this information about the default version from the subsystem help string and have it just in places that don't cause the subclassing fiddle, i.e. option help text like `install_from_resolve`.
   1. Variant: omit it _for now_ so we can land this PR with just the parsing & display in option text and no change to the top-level help text, and then consider how to display it elsewhere as follow-up 

For 2, If we were to add a new field, I think that'd mean:

- adding an field `help_short` (or some other name) to `PythonToolRequirementBase`
- updating the ~40 subclasses to swap `help` -> <new name>, which would be annoying but not horrible
- finessing how `PythonToolRequirementBase` sets its `help` property to the extra-formatting version, to work better with mypy + existing classes (I think it'd have to be `help: ClassVar[str | Callable[[], str] = _real_help_classmethod` with the type annotation)
- communicating the change for plugins in release notes, something like "subclasses of `PythonToolRequirementBase` and `PythonToolBase` can replace `help` with <new name> to use the default help behaviour which inserts the default tool version".

For 3, that might be something like a class method `def default_version_help_text(cls) -> str` and downstream classes like `Black` & `ClangFormat` etc. have to call it in their help constructors f"... {cls.default_version_help_text()}"`.

I'm somewhat inclined to 4i (i.e. split this PR up so that there's a separate PR with the change to the top-level help text), so that this PR isn't blocked in this discussion and we get to see some of the benefits of your work sooner. But, I'm unsure.

What do you think @krishnan-chandra?

Reply:
I agree with following 4i for now, and splitting out the changes to the top-level help text into a separate PR. 

On a go-forward basis, here are my thoughts:

The current content of the PR is basically the same as approach 1 -

> 1. keep this approach, with a new method on the Subsystem class which lets "leaf" subclasses just set help while still giving superclasses flexibility for manipulating it

I understand the concern around the new method being public - I think making this method private could potentially help, although we'd have to call `getattr` on a private attribute which feels a bit icky.

> 2. a new field manipulated by the Python tool helpers, that Python tool subclasses need to set

This one might be best, especially if we make `help_short` required? That would give plugin developers a clear indicator of what to set, and would also let us catch unset values via type-checking.

> 3. a class method that Python tool subclasses can call when formatting their own help text
> For 3, that might be something like a class method def default_version_help_text(cls) -> str and downstream classes like Black & ClangFormat etc. have to call it in their help constructors f"... {cls.default_version_help_text()}"`.

While this could work, I think there's some footguns here - if you're making a new Python tool, you could very easily end up calling `.help` instead of `.default_version_help_text` or miss the call entirely, which seems suboptimal.

On the whole, I'd probably vote for option 2 out of all the ones you proposed.